content/posts/2017/2017-10-17-plotly_vs_ggplotly.Rmd

```diff
@@ -0,0 +1,118 @@
+
+### Prerequisites 
+
+Throughout we will be using the packages: __dplyr__, __ggplot2__, __plotly__ and __microbenchmark__. The data in use is the `birthdays` dataset in the __mosaicData__ package. Containing the counts of births in USA on each day from 1969 - 1988.
+
+```{r, eval = FALSE, message=FALSE, warning=FALSE}
+install.packages(c("mosaicData", "dplyr", "tidyr","ggplot2","plotly", "microbenchmark"))
+```
+``` {r, warning =FALSE, message = FALSE, error = FALSE}
+library(mosaicData)
+library(dplyr)
+library(tidyr)
+library(ggplot2)
+library(plotly)
+library(microbenchmark)
+```
+
+### Analysis
+Let's loads and take a look at the data.
+```{r}
+data("Birthdays", package = "mosaicData")
+head(Birthdays)
+```
+
+Let's start off with a very simple scatter graph of the mean births in every year. 
+
+```{r}
+meanb = Birthdays %>% 
+  group_by(year) %>% 
+     summarise(mean = mean(births))
+```
+
+```{r, warning = FALSE, error=FALSE, message=FALSE}
+ggplotly(ggplot(meanb) + 
+  geom_point(aes(y = mean, x = year, colour = year)))
+
+plot_ly(data = meanb, 
+                 y = ~mean, x = ~year, color = ~year, 
+                 type = "scatter")
+```
+
+Both graphs identical bar styling, yes? 
+
+Now let's use `microbenchmark` to see how their timings compare. 
+
+```{r, warning = FALSE, error=FALSE, message=FALSE}
+time = microbenchmark::microbenchmark(
+        ggplotly = ggplotly(ggplot(meanb) + 
+                            geom_point(aes(y = mean, x = year, colour = year))),
+          plotly = plot_ly(data = meanb, 
+                           y = ~mean, x = ~year, 
+                           color = ~year, type = "scatter"),
+                           times = 100, unit = "s")
+time
+```
+```{r, warning = FALSE, error=FALSE, message=FALSE}
+autoplot(time)
+```
+
+```{r, include=FALSE}
+timetable = cbind(expr = time$expr, time = time$time)
+mean = as.data.frame(timetable) %>% group_by(expr) %>% summarise(mean = mean(time))
+average = round(mean[1,2]/mean[2,2], 0)
+```
+Now I thought nesting a __ggplot__ object within `ggplotly()` would be slower than using `plot_ly()`, but I didn't think it would be this slow. On average `ggplotly()` is `r average` times slower than `plot_ly()`. **_`r average`!_** One run even took `r round(microbenchmark:::convert_to_unit(max(time$time), unit = "s"), 3)` seconds! 
+
+Let's take it up a notch.  There we were just plotting 20 points, what about if we plot over 20,000? Here we will plot the min, mean and max births on each day. 
+
+```{r warning = FALSE, error=FALSE, message=FALSE}
+date = Birthdays %>% 
+  group_by(date) %>% 
+  summarise(mean = mean(births), min = max(births), max = min(births))
+all = date %>% 
```

Comment:
Why not just pipe summerise into gather?

Reply:
long day... of course!